### ISSUE TITLE
🚀 Feature: Azure Blob Storage entity provider

### ISSUE THREAD

**Issue by drodil** [backstage/backstage#23130]
### 🔖 Feature description

Just like AWS S3 Entity Provider, there could be support to fetch entities from Azure Blob Storage

### 🎤 Context

Azure users cannot use Blob Storage to serve catalog info files for Backstage like AWS S3 users can. A new provider would be needed to support this.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by tudi2d** [comment]
Thanks @drodil - sounds good! Would you be interested in contributing the provider or should we open this up for contributions?

**Comment by numcys** [comment]
hey @drodil I am a new contributor here, but have experience with cloud, especially azure. If possible I would like to work on this one, if you you would be willing to point me to the currently implemented AWS S3 provider...

**Comment by drodil** [comment]
Hey @numcys , that would br great! The S3 provider is located here: https://github.com/backstage/backstage/blob/master/plugins%2Fcatalog-backend-module-aws%2Fsrc%2Fproviders%2FAwsS3EntityProvider.ts

**Comment by drodil** [comment]
Oh, and @numcys , this would probably be the correct module for that: https://github.com/backstage/backstage/tree/master/plugins/catalog-backend-module-azure

**Comment by numcys** [comment]
Hi @drodil, I have been tackling some problems while trying to implement this, I didn't had much experience with typescript and even though I tried, I don't think I will be able to complete this one. Really sorry if I wasted yours or anyone's time and really thankful for giving me this chance. I will take my time to understand the project better and then try to go for contributions...

**Comment by Nikunj0601** [comment]
Hi @drodil @freben If anyone is not working on this I would like to work on this one. Also, you want separate file for azure blob entity provider or you want add to azure devops file?

**Comment by drodil** [comment]
@Nikunj0601 no I don't think there's anyone working on this yet so you are welcome to take it 👌 What do you think is the best way to proceed?

**Comment by Nikunj0601** [comment]
I think it should be in separate cause in catalog provider accountName, accountKey, contiainerKey which is from devOps provider. Though I am not sure it is a best approach or not.

**Comment by Nikunj0601** [comment]
Hi @drodil @freben I was trying to add aws s3 plugin but it seems that i am doing something wrong while integrating it. I followed the backstage docs https://backstage.io/docs/integrations/aws-s3/discovery. Also I have completed entityProvider and integration can you take a look and give a feedback whether I am working on right direction or not.
https://github.com/Nikunj0601/backstage/tree/feat/azure-blob-storage-integration

**Comment by Nikunj0601** [comment]
@drodil @freben any idea why log says one location committed but component is not visible in ui?
backend log: 
`[example-backend]: 2024-05-19T22:13:39.371Z catalog info Committed 1 Locations for Azure Blob Storage blobs target=azureBlobStorage-provider:default class=AzureBlobStorageEntityProvider taskId=azureBlobStorage-provider:default:refresh taskInstanceId=5f2b1fae-cc04-44bb-839f-41a9c6d71669`
Also, is there a cmd to create new plugin was added in `azureDevOps` but allows only one module which is set to export as `default`. Sorry to bug you on weekend.

**Comment by Nikunj0601** [comment]
@Rugvip Thanks I will do that for azure blob storage. Also, is there any cmd to create a new plugin cause intially I added this azure blob entity provider in azure DevOps plugin but while exporting `catalogModuleAzureBlobStorageEntityProvider` in `/module/index.ts`. for entityProvider to be recoginsed it needs to be exported as default but there is already azureDevopsEntityProvider exported as default. So I think creating new plugin would be better. And one more question this storage account should always be public or it can also be private.

**Comment by Nikunj0601** [comment]
@Rugvip I am having an issue with registering AzureBlobStorageUrlReader, I have added factory in [UrlReaders](https://github.com/backstage/backstage/blob/master/packages/backend-common/src/reading/UrlReaders.ts) also exported AzureBlobStorageUrlReader in index.ts as well but I can see any logs or signs that being actually used. Can you guide me I like am I missing export it somewhere as well? Thanks

**Comment by freben** [comment]
Shouldn't be about a missing export if typescript is happy and not complaining about your code, I think.

Is this on the new backend system? Did you remember to register a replacement `coreServices.urlReader` in your backend with these additions in it?

Also, what did you do to try to trigger it being called? Some URL registered in the catalog or so?

**Comment by Nikunj0601** [comment]
@freben Yes I did registered a factory in `UrlReaders` class. Also, for triggerring I am adding AzureBlobStorageEntityProvider which I created into `catalog.ts` using 
`builder.addEntityProvider(
    ...AzureBlobStorageEntityProvider.fromConfig(env.config, {
      logger: env.logger,
      schedule: env.scheduler.createScheduledTaskRunner({
        frequency: Duration.fromObject({ minutes: 30 }),
        timeout: Duration.fromObject({ minutes: 3 }),
      }),
    }),
  );` 
  which is similar to importing awsS3 entityProvider.

**Comment by drodil** [comment]
@Nikunj0601 any updates for this?

**Comment by Nikunj0601** [comment]
@drodil It is almost done but I haven't got much time in last month because of studies but I will get it done as I will finish my masters in few weeks. Just writing testcases and few nitpicks are remaining.

**Comment by Nikunj0601** [comment]
@drodil @freben Sorry I got little bit busy with academics so could complete on time. I have added the entity provider for azure blob storage in catalog-backend-module-azure/module but while exporting it in index.js 
`
export { catalogModuleAzureDevOpsEntityProvider } from './catalogModuleAzureDevOpsEntityProvider';
export { catalogModuleAzureBlobStorageEntityProvider as default } from './catalogModuleAzureBlobStorageEntityProvider';
`
If it export it as default then it is working properly but if I remove default it is give me error as below from the `BackstageBackend.ts`
`
node:79597) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 message listeners added to [process]. Use emitter.setMaxListeners() to increase limit
[example-backend]: /Users/nikunjhudka/Documents/dalhousie/Nikunj personal projects/backstage/packages/backend-app-api/src/wiring/BackstageBackend.ts:68
[example-backend]:   if ('$$type' in feature) {
[example-backend]:       ^
[example-backend]: 
[example-backend]: 
[example-backend]: TypeError: Cannot use 'in' operator to search for '$$type' in undefined
[example-backend]:     at unwrapFeature (/Users/nikunjhudka/Documents/dalhousie/Nikunj personal projects/backstage/packages/backend-app-api/src/wiring/BackstageBackend.ts:68:7)
[example-backend]:     at f (/Users/nikunjhudka/Documents/dalhousie/Nikunj personal projects/backstage/packages/backend-app-api/src/wiring/BackstageBackend.ts:35:47)
[example-backend]:     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[example-backend]: 
[example-backend]: Node.js v20.12.2
`

Should I create the different catalog-backend-module or is there a way to resolve this? If not then is there a command to create one with boiler plate code.

Thank you

**Comment by freben** [comment]
Yeah it needs to be a default export; that's how we set up the contract for these.

So we have two options:

- Make a new npm package, just to hold this provider
- Make a single module as the default export of the existing package, and that module (in its `registerInit`) adds TWO providers in one go.

I'm voting for the second one. Both. providers would be activated by the presence or absence of the corresponding config, so it's fine that we stay "less granular" (adding many things with one module), because there's not really any waste or performance cost to worry about or so.

**Comment by Nikunj0601** [comment]
@drodil @freben Just raised the PR for the feature. Looking forward to the review and feedback. And, sorry for taking such a long time I just got bit busy with school work.

**Comment by RohitThomas-PI** [comment]
@Nikunj0601 I just used the plugin that you created and it seems to read it from catalog-import but not from the catalog discovery any idea why ?
this is the error that I see 
`catalog warn Unable to read url, RestError: Could not retrieve file from Azure Blob Storage; caused by RestError entity=location:default/generated-ab74f52f35744da1f8c7e25d90ca553dbe4e7edc location=url:https://***.blob.core.windows.net/***/catalog/catalog-info.yaml`

Also I noticed that I am see an error being generated from the locations for the catalog I imported from.
![image](https://github.com/user-attachments/assets/107d7a04-1445-4b8d-93db-84c87d49205e)
But I can still update the catalog and it will get reflected if I do a refresh(the error temporarily disappears)

**Comment by Nikunj0601** [comment]
> @Nikunj0601 I just used the plugin that you created and it seems to read it from catalog-import but not from the catalog discovery any idea why ? this is the error that I see `catalog warn Unable to read url, RestError: Could not retrieve file from Azure Blob Storage; caused by RestError entity=location:default/generated-ab74f52f35744da1f8c7e25d90ca553dbe4e7edc location=url:https://***.blob.core.windows.net/***/catalog/catalog-info.yaml`
> 
> Also I noticed that I am see an error being generated from the locations for the catalog I imported from. ![image](https://private-user-images.githubusercontent.com/127098294/391589149-107d7a04-1445-4b8d-93db-84c87d49205e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMxNTE2NDUsIm5iZiI6MTczMzE1MTM0NSwicGF0aCI6Ii8xMjcwOTgyOTQvMzkxNTg5MTQ5LTEwN2Q3YTA0LTE0NDUtNGI4ZC05M2RiLTg0Yzg3ZDQ5MjA1ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMlQxNDU1NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OWEzYjA2NTM5ZDg3NTUyZGQwOTBmYmM3NGUzOWM1ZmI0MWE3ZThiYzhhYzFhZTAwYjFmYTBiNzRmODQyMzYwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ij6MYZeczuc6KH4_fwMPCxJy3faS5KTgP7S3Tftqh9E) But I can still update the catalog and it will get reflected if I do a refresh(the error temporarily disappears)

@RohitThomas-PI I just checked it's working for i am not getting any error. Can please provide how did you configured the plugin in `app-config.yaml`.

**Comment by RohitThomas-PI** [comment]
I have added it based on what you have mentioned in the docs:

```
integrations:
  azureBlobStorage:
    - accountName: storageName # required
      aadCredential:
        clientId: ${AZURE_CLIENT_ID}
        tenantId: ${AZURE_TENANT_ID}
        clientSecret: ${AZURE_CLIENT_SECRET}

catalog:
  providers:
    azureBlob:
      accountName: storageName
      containerName: containerName
      schedule: # same options as in TaskScheduleDefinition
        # supports cron, ISO duration, "human duration" as used in code
        frequency: { minutes: 3 }
        # supports ISO duration, "human duration" as used in code
        timeout: { minutes: 3 }
```
I had to skip adding the endpoint as that didnt work.

I have given the following IAM access roles(storage account ) for backstage

- Storage Blob Data Owner 
- Storage Queue Data Contributor

**Comment by RohitThomas-PI** [comment]
That's great to see that its a simple fix, any idea about the catalog discovery not being able to fetch ? is it a permission issue ? or a configuration issue ?

**Comment by Nikunj0601** [comment]
@azizabah I believe it is because you need to add the catalog configuration as well.
```
azureBlob:
      provider1: 
        accountName: myAccount
        containerName: myContainer
        schedule: # same options as in TaskScheduleDefinition
          # supports cron, ISO duration, "human duration" as used in code
          frequency: { minutes: 30 }
          # supports ISO duration, "human duration" as used in code
          timeout: { minutes: 3 }
```

if still it does not resolve the issue then please create a separate issue and i will take a look.

**Comment by azizabah** [comment]
Adding the catalog config leads to a new error where it is now iterating over every item inside the storageaccount/container but it is throwing a 404 for each one (which is very interesting)
```
[backend]: 2025-01-07T16:58:32.881Z catalog warn Unable to read url, NotFoundError: could not read https://***.blob.core.windows.net/myContainer/service1/openapi.json, 404 The specified resource does not exist. entity=location:default/generated-be87990dcd44a269882c2a1d8e43ad7498113a85 location=url:https://***.blob.core.windows.net/myContainer/service1/openapi.json

[backend]: 2025-01-07T16:58:32.886Z catalog warn Unable to read url, NotFoundError: could not read https://***.blob.core.windows.net/myContainer/service2/openapi.json, 404 The specified resource does not exist. entity=location:default/generated-341fd20d7102e6b7257040d884d6870d0497a3ec location=url:https://***.blob.core.windows.net/myContainer/service2/openapi.json
```

The service principal that I'm using to integrate has the following permissions on the storage account:
"Storage Blob Data Owner"
"Storage Blob Data Contributor"

The 404 makes me think this that something is taking the correct URLs that it has found and not passing the correct auth on retrieving them.

**Comment by Nikunj0601** [comment]
@azizabah I tried reproducing the issue but it is working for me so looks like azure active directory configuration issue or storage account issue. I have created a account and put a api entity and open-api yaml for defination and it is working for me.

backstage config is same as you provided.

![image](https://github.com/user-attachments/assets/640e81dc-5d6c-428b-b31d-6100d35dd62f)
![image](https://github.com/user-attachments/assets/e584a67e-538f-4a64-9be6-fbf7acd4e8e8)
![image](https://github.com/user-attachments/assets/76005cd1-4c07-488c-bf88-c12ad100e886)

**Comment by azizabah** [comment]
404 is the error you receive if you go to the correct URL without authentication. If you copy the URL into a browser and do a simple GET you'd see a 404.

**Comment by Nikunj0601** [comment]
when i copied the url and tried to GET in browser i am getting PermissionDenied error. can you please confirm the url from azure control panel.
![image](https://github.com/user-attachments/assets/0d0d119c-7d58-45a0-a9cb-3fb5e2743d60)

**Comment by azizabah** [comment]
I'm just going to the blob and copy+pasting that URL to get a URL of format 
https://storageAccountName.blob.core.windows.net/container/folder/openApi.json

When pasting into Chrome, I get 
<Error>
<Code>ResourceNotFound</Code>
<Message>The specified resource does not exist. RequestId:b79c7ac3-e01e-0042-3846-61d541000000 Time:2025-01-07T20:56:48.9127137Z</Message>
</Error>

**Comment by Ishujeet** [comment]
@azizabah @Nikunj0601  guys, any update on this 404 issue. I have tried all 3 methods for authentication, when using SAS it throws invalid authentication headers, and for other two is showing 404.

**Comment by Ishujeet** [comment]
@azizabah  found the fix, it is due to access level 

<img width="524" alt="Image" src="https://github.com/user-attachments/assets/32285a2d-db68-4d32-98d1-727a5530dd64" />
